libsignal-protocol/src/macros.rs

```diff
@@ -37,3 +37,53 @@ macro_rules! impl_is_a {
         )*
     };
 }
+
+/// An an alternative to [`assert!()`] for functions called from C.
+macro_rules! signal_assert {
+    ($condition:expr) => {
+        if !$condition {
+            ::log::error!(
+                "Assertion failed at {}#{}: {}",
+                file!(),
+                line!(),
+                stringify!($condition),
+            );
+            ::log::error!("{}", ::failure::Backtrace::new());
+
+            eprintln!(
+                "Aborting because the application reached an unexpected state at {}#{}: {}",
+                file!(),
+                line!(),
+                stringify!($condition)
+            );
+            ::std::process::exit(1);
```

Comment:
Hmmm, i think it would be better to let that option to the consumer, maybe a simple feature flag `abort-on-ub` enabled by default, so if it's not enabled just return the `InternalError::Unknown`

Reply:
Looking back at where this macro gets used, it seems it's almost always used to validate arguments at the start of an `unsafe extern "C"` function, so why not **always** return an `InternalError::Argument`?

Originally I was tempted to say the world is in an indeterminate state if we ever get a null pointer and we should crash loudly, but on the other hand, it  could be thought of as just another error due to dodgy input which we can recover from (e.g. by returning `Result<T, E>` or an error code).

Reply:
Yeah, i noticed that too, and i don't think it's very bad experience to just crash the whole world, i know it would be better in some cases, but we should better leave that decision for the consumer of the library .